File: Days 51 - 60/Day 53/QueueWithTwoStacks.cpp
#include <iostream>
#include <stack>
#include <stdexcept>

class Queue
{
private:
	std::stack<int> stackA;
	std::stack<int> stackB;

public:
	void Enqueue(int value)
	{
		stackA.push(value);
	}

	void Dequeue()
	{
		if (stackA.empty())
		{
			throw std::out_of_range("Queue is empty.");
		}

		SwapStacks(stackA, stackB);
		stackB.pop();

		SwapStacks(stackB, stackA);
	}

	int Peek()
	{
		if (stackA.empty())
		{
			throw std::out_of_range("Queue is empty.");
		}

		SwapStacks(stackA, stackB);

		int value = stackB.top();

		SwapStacks(stackB, stackA);

		return value;
	}

private:
	void SwapStacks(std::stack<int>& a, std::stack<int>& b)
	{
		while (!a.empty())
		{
			b.push(a.top());
			a.pop();
		}
	}
};

int main(int argc, const char* argv[])
{
	Queue queue;

	queue.Enqueue(5);
	queue.Enqueue(2);
	queue.Enqueue(7);
	std::cout << queue.Peek() << "\n";

	queue.Dequeue();
	std::cout << queue.Peek() << "\n";

	std::cin.get();

	return 0;
}

File: Days 051 - 060/Day 53/QueueWithTwoStacks.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Queue
	{
		private Stack<int> stackA = new Stack<int>();
		private Stack<int> stackB = new Stack<int>();

		public void Enqueue(int item)
		{
			stackA.Push(item);
		}

		public void Dequeue()
		{
			if (stackA.Count == 0)
			{
				throw new InvalidOperationException("Queue is empty.");
			}

			SwapStacks(stackA, stackB);
			stackB.Pop();
			SwapStacks(stackB, stackA);
		}

		public int Peek()
		{
			if (stackA.Count == 0)
			{
				throw new InvalidOperationException("Queue is empty.");
			}

			SwapStacks(stackA, stackB);

			int item = stackB.Peek();

			SwapStacks(stackB, stackA);

			return item;
		}

		private void SwapStacks(Stack<int> a, Stack<int> b)
		{
			while (a.Count > 0)
			{
				b.Push(a.Peek());
				a.Pop();
			}
		}
	}

	internal class Day53
	{
		private static int Main(string[] args)
		{
			Queue queue = new Queue();
			queue.Enqueue(5);
			queue.Enqueue(2);
			queue.Enqueue(7);
			Console.WriteLine(queue.Peek());

			queue.Dequeue();
			Console.WriteLine(queue.Peek());

			Console.ReadLine();

			return 0;
		}
	}
}